### PR TITLE
Create API to view & update user's personal details

### PR DESCRIPTION
**Pull request by mohinid**
## Notion card
https://www.notion.so/saeloun/Create-APIs-to-add-and-view-personal-details-for-a-user-e312e4f0ceb849fda5f14eb05f8b2cf3

## Summary
1. Create an API for updating and viewing users' personal details from users table.

## Type of change
- [x] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to
      not work as expected)
- [x] This change requires a documentation update

## How Has This Been Tested?
Used Postman on local & RSpecs.

### Checklist:

- [x] I have manually tested all workflows
- [x] I have performed a self-review of my own code
- [x] I have added automated tests for my code

### CODE REVIEW

**Comment by mohinid** [comment]
Postman response for User Details API
Screenshot 2022-07-05 at 7 03 52 PM

File: app/controllers/internal_api/v1/team_members/details_controller.rb [hunk 1]
```diff
@@ -0,0 +1,10 @@
+# frozen_string_literal: true
+
+class InternalApi::V1::TeamMembers::DetailsController < InternalApi::V1::ApplicationController
+  def show
+    authorize current_user, policy_class: TeamMembers::DetailPolicy
```
Comment (keshavbiswa) [inline app/controllers/internal_api/v1/team_members/details_controller.rb:5]: You need to authorize the resource, not the current_user. In this action the resource would be the user.

File: app/views/internal_api/v1/team_members/details/show.json.jbuilder [hunk 2]
```diff
@@ -0,0 +1,11 @@
+# frozen_string_literal: true
+
+json.details do
+  json.first_name user.first_name
+  json.last_name user.last_name
+  json.date_of_birth user.date_of_birth
+  json.phone user.phone
+  json.personal_email_id user.personal_email_id
+  json.github_url user.social_accounts["github_url"]
+  json.linkedin_url user.social_accounts["linkedin_url"]
```
Comment (keshavbiswa) [inline app/views/internal_api/v1/team_members/details/show.json.jbuilder:10]: You can make use of the `extract` method.

File: app/policies/team_members/detail_policy.rb [hunk 3]
```diff
@@ -0,0 +1,7 @@
+# frozen_string_literal: true
+
+class TeamMembers::DetailPolicy < ApplicationPolicy
+  def show?
+    user_owner_role? || user_admin_role? || user_employee_role?
```
Comment (keshavbiswa) [inline app/policies/team_members/detail_policy.rb:5]: Take a look at `.has_any_role?` from the `pundit` gem documentation.

**Review by keshavbiswa — CHANGES_REQUESTED** [review]

File: app/controllers/internal_api/v1/team_members/details_controller.rb [hunk 4]
```diff
@@ -0,0 +1,32 @@
...
+    authorize employment, policy_class: TeamMembers::DetailPolicy
+    user = employment.user
+    render :show, locals: { user: }, status: :ok
+  end
+
+  def update
+    employment = Employment.find(params[:team_id])
+    authorize employment, policy_class: TeamMembers::DetailPolicy
+    user = employment.user
+    if user.update!(detail_params)
```
[the remark this answers is not in the document]
Comment (mohinid) [reply app/controllers/internal_api/v1/team_members/details_controller.rb:15]: removed if condition

**Comment by mohinid** [comment]
> Should someone with admin or employee role, be able to update any team member's personal details?

If from the same company- yes. @supriya3105  please confirm.

**Comment by supriya3105** [comment]
> > Should someone with admin or employee role, be able to update any team member's personal details?
> 
> If from the same company, then yes. @supriya3105 please confirm.

Yes. I agree . Admin and owner should be able to update personal details of any employee

File: app/controllers/internal_api/v1/team_members/details_controller.rb [hunk 5]
```diff
@@ -0,0 +1,26 @@
...
+  def update
+    employment = Employment.find(params[:team_id])
+    authorize employment, policy_class: TeamMembers::DetailPolicy
+    user = employment.user
+    user.update!(detail_params)
+    render :update, locals: { user: }, status: :ok
+  end
+
+  private
+
```
Comment (akhilgkrishnan) [inline app/controllers/internal_api/v1/team_members/details_controller.rb:19]: You can use memoization for find the employement record.
def employment
 @employment ||= Employment.find(params[:team_id])
end

File: app/controllers/internal_api/v1/team_members/details_controller.rb [hunk 6]
```diff
@@ -0,0 +1,26 @@
+# frozen_string_literal: true
+
+class InternalApi::V1::TeamMembers::DetailsController < InternalApi::V1::ApplicationController
+  def show
+    employment = Employment.find(params[:team_id])
+    authorize employment, policy_class: TeamMembers::DetailPolicy
+    render :show, locals: { user: employment.user }, status: :ok
```
Comment (akhilgkrishnan) [inline app/controllers/internal_api/v1/team_members/details_controller.rb:7]: include proper message to the response

**Review by akhilgkrishnan — COMMENTED** [review]
Added couple of comments.

File: app/policies/team_members/detail_policy.rb [hunk 7]
```diff
@@ -0,0 +1,15 @@
...
+  def show?
+    user.has_any_role?(
+      { name: :admin, resource: record.company },
+      { name: :owner, resource: record.company }) || user == record.user
+  end
+
+  def update?
+    user.has_any_role?(
+      { name: :admin, resource: record.company },
+      { name: :employee, resource: record.company }) || user == record.user
```
Comment (akhilgkrishnan) [inline app/policies/team_members/detail_policy.rb:13]:
```suggestion
      { name: :owner, resource: record.company }) || user == record.user
```

Comment (mohinid) [reply app/policies/team_members/detail_policy.rb:13]: my bad! :/

**Review by akhilgkrishnan — APPROVED** [review]
LGTM 🎉

File: spec/factories/users.rb [hunk 8]
```diff
@@ -7,6 +7,9 @@
     email { Faker::Internet.safe_email }
     password { Faker::Internet.password }
     confirmed_at { Date.today }
+    date_of_birth { Date.today }
```
Comment (keshavbiswa) [inline spec/factories/users.rb:10]:
```suggestion
    date_of_birth { Date.current }
```

File: spec/requests/internal_api/v1/team_members/details/show_spec.rb [hunk 9]
```diff
@@ -0,0 +1,108 @@
...
+    it "is successful" do
+      expect(response).to have_http_status(:ok)
+    end
+  end
+
+  context "when Admin wants to see details of employee of his company" do
+    before do
+      user.add_role :admin, company
+      sign_in user
+      get "/internal_api/v1/team/#{employment.id}/details"
```
Comment (keshavbiswa) [inline spec/requests/internal_api/v1/team_members/details/show_spec.rb:28]: You can make use of the `send_request` method.

**Comment by mohinid** [comment]
@keshavbiswa integrated suggested changes. Ci is failing for some other test. Will merge it once resolved.

File: spec/requests/internal_api/v1/team_members/details/show_spec.rb [hunk 10]
```diff
@@ -0,0 +1,108 @@
...
+
+  context "when logged in user checks his own details" do
+    before do
+      user.add_role :employee, company
+      sign_in user
+      send_request :get, internal_api_v1_team_details_path(employment)
+    end
+
+    it "is successful" do
+      expect(response).to have_http_status(:ok)
```
Comment (keshavbiswa) [inline spec/requests/internal_api/v1/team_members/details/show_spec.rb:44]: Could you also assert if we're getting the correct response data?

File: spec/requests/internal_api/v1/team_members/details/show_spec.rb [hunk 11]
```diff
@@ -0,0 +1,108 @@
...
+
+  context "when Owner wants to see details of employee of his company" do
+    before do
+      user.add_role :owner, company
+      sign_in user
+      send_request :get, internal_api_v1_team_details_path(employment)
+    end
+
+    it "is successful" do
+      expect(response).to have_http_status(:ok)
```
Comment (keshavbiswa) [inline spec/requests/internal_api/v1/team_members/details/show_spec.rb:20]: Could you also assert if we're getting the correct response data?

File: spec/requests/internal_api/v1/team_members/details/show_spec.rb [hunk 12]
```diff
@@ -0,0 +1,108 @@
...
+
+  context "when Admin wants to see details of employee of his company" do
+    before do
+      user.add_role :admin, company
+      sign_in user
+      send_request :get, internal_api_v1_team_details_path(employment)
+    end
+
+    it "is successful" do
+      expect(response).to have_http_status(:ok)
```
Comment (keshavbiswa) [inline spec/requests/internal_api/v1/team_members/details/show_spec.rb:32]: Could you also assert if we're getting the correct response data?

File: spec/requests/internal_api/v1/team_members/details/update_spec.rb [hunk 13]
```diff
@@ -0,0 +1,153 @@
...
+      send_request :patch, internal_api_v1_team_details_path(
+        team_id: employment.id,
+        params: {
+          user: @user_details
+        })
+    end
+
+    it "is successful" do
+      user.reload
+      expect(response).to have_http_status(:ok)
```
Comment (keshavbiswa) [inline spec/requests/internal_api/v1/team_members/details/update_spec.rb:57]: Could you also assert if we're getting the correct response data?

File: spec/requests/internal_api/v1/team_members/details/update_spec.rb [hunk 14]
```diff
@@ -0,0 +1,153 @@
...
```
Comment (keshavbiswa) [inline spec/requests/internal_api/v1/team_members/details/update_spec.rb:74]: Could you also assert if we're getting the correct response data?

File: spec/requests/internal_api/v1/team_members/details/show_spec.rb [hunk 15]
```diff
@@ -18,6 +18,7 @@
 
     it "is successful" do
       expect(response).to have_http_status(:ok)
+      response.body == user.to_json
```
Comment (keshavbiswa) [inline spec/requests/internal_api/v1/team_members/details/show_spec.rb:21]: This is not a correct way to write tests.

**Review by keshavbiswa — CHANGES_REQUESTED** [review]

File: spec/requests/internal_api/v1/team_members/details/show_spec.rb (hunk 15 above)
Comment (mohinid) [reply spec/requests/internal_api/v1/team_members/details/show_spec.rb]: please check now.

File: spec/requests/internal_api/v1/team_members/details/show_spec.rb [hunk 16]
```diff
@@ -18,7 +18,12 @@
 
     it "is successful" do
       expect(response).to have_http_status(:ok)
-      response.body == user.to_json
+      expect(json_response["first_name"]).to eq(JSON.parse(user.first_name.to_json))
+      expect(json_response["last_name"]).to eq(JSON.parse(user.last_name.to_json))
+      expect(json_response["personal_email_id"]).to eq(JSON.parse(user.personal_email_id.to_json))
+      expect(json_response["date_of_birth"]).to eq(JSON.parse(user.date_of_birth.to_json))
+      expect(json_response["phone"]).to eq(JSON.parse(user.phone.to_json))
+      expect(json_response["social_accounts"]).to eq(JSON.parse(user.social_accounts.to_json))
```
Comment (keshavbiswa) [inline spec/requests/internal_api/v1/team_members/details/show_spec.rb:26]: expect(json_response).to eq(user.to_json) this could also work.

Comment (mohinid) [reply spec/requests/internal_api/v1/team_members/details/show_spec.rb:26]: its not working because we have checks on specific fields.

Comment (mohinid) [reply spec/requests/internal_api/v1/team_members/details/show_spec.rb:26]: Here is the error SS- 
Screenshot 2022-07-12 at 1 18 50 PM

**Pull request merged**